### DISCUSSION TITLE
What is the correct way of passing multiple variable arguments to a riverpod provider

### DISCUSSION THREAD

**Discussion by sprabhdeep** [rrousselGit/riverpod#3517] [Q&A]
I am passing arguments to my provider like

```
@riverpod
class FetchProducts extends _$FetchProducts {
  @override
  Future<List<Product>> build({
    required String? specificProductId,
    required String? clientUid,
  }) async {
    List<Product> _productList = [];

    if (specificProductId != null) {
      final productDataSnapshot = await FirebaseFirestore.instance
          .collection('product')
          .doc(specificProductId.trim())
          .get();
      if (productDataSnapshot.exists && productDataSnapshot.data() != null) {
        _productList.add(Product.fromMap(productDataSnapshot.data()!));
      }
    } else if (clientUid != null && clientUid != '') {
      final productDataSnapshot = await FirebaseFirestore.instance
          .collection('product')
          .where('client_uid', isEqualTo: clientUid)
          .get();
      print(productDataSnapshot.docs.map((data) => data.data()));
      for (var pdt in productDataSnapshot.docs) {
        _productList.add(Product.fromMap(pdt.data()));
      }
    } else {
      final productDataSnapshot = await FirebaseFirestore.instance.collection('product').get();
      for (var pdt in productDataSnapshot.docs) {
        _productList.add(Product.fromMap(pdt.data()));
      }
    }

    return _productList;
  }

  Future<void> createNewProduct(Product pdtData) async {
    final newPdt = await FirebaseFirestore.instance.collection('product').add(pdtData.toMap());
    Product newProduct = Product.fromMap(newPdt as Map<String, dynamic>);
    state = AsyncData([...?state.value, newProduct]);
  }
}
```

As soon as my UI page opens all products are fetched by a widget with both specificProductId and clientUid as null/empty.

`final productList = ref.watch(fetchProductsProvider(clientUid: null, specificProductId: null));`

There is another notifier provider which gives clientUid when a client is chosen in UI. This triggers a widget on the same UI page to fetch products for that specific clientUid. `ref.watch(FetchProductsProvider(specificProductId: null, clientUid: widget.clientUid));`

However, this causes the provider with all the products i.e. `productList` getting disposed off.

Problem: How to keep the state of the provider with all the products intact when provider is watched by any other widget with different set of parameters, but auto-disposing the provider when the last listener has been removed?

Current Console output looks like -

> Provider fetchProductsProvider:FetchProductsProvider#80775(null) was initialized with
> AsyncLoading<List<Product>>()
> Provider fetchProductsProvider:FetchProductsProvider#80775(null) updated from
> AsyncLoading<List<Product>>() to AsyncData<List<Product>>(value: [Instance of 'Product',
> Instance of 'Product', Instance of 'Product', Instance of 'Product', Instance of 'Product'])
> 
> // Chosing a client
> 
> Provider chosenClientUidProvider:AutoDisposeNotifierProviderImpl<ChosenClientUid,
> String?>#d6c1c updated from  to 8C9rnBB2sNk0
> Provider fetchProductsProvider:FetchProductsProvider#4e744(null) was initialized with
> AsyncLoading<List<Product>>()
> Provider fetchProductsProvider:FetchProductsProvider#80775(null) was disposed
> Provider fetchProductsProvider:FetchProductsProvider#4e744(null) updated from
> AsyncLoading<List<Product>>() to AsyncData<List<Product>>(value: [Instance of 'Product',
> Instance of 'Product', Instance of 'Product'])

**Comment by gdurandrexel** [comment]
You could create an AllProducts provider that fetches all your product (using the code of your current provider, but with no parameter) and then have the provider FetchProducts with the parameters depending on AllProducts and filtering using the parameters.

**Reply by sprabhdeep** [reply]
Thanks. Creating two providers might solve but may introduce inefficiency in number of database reads. Is there any why to list all alive providers at certain point in my build function?
Asking because I feel riverpod is initializing a provider twice with same # value. For example, in my case
1. First initialization for `final productList = ref.watch(fetchProductsProvider(clientUid: null, specificProductId: null));`
2. Second initialization with same # 80775 value in another widget on same UI page `ref.watch(FetchProductsProvider(specificProductId: null, clientUid: widget.clientUid));` because widget.clientUid is empty ('') initially

And when I chose a client, a provider is disposed but it may be the case another initilization in the first widget is still alive? Any method to check that?

**Reply by sprabhdeep** [reply]
I tested using below code in the build method of my second widget
```
    print('All products list provider exists?: ${ref.exists(fetchProductsProvider(specificProductId: null, clientUid: null))}');
    print('Client specific products list provider exists?: ${ref.exists(fetchProductsProvider(specificProductId: null, clientUid: ''))}');
```
I get `true` for both prints before choosing any client and `true` & `false` respectively in that order after I choose a client from dropdown in my second widget. 

I can only be sure until I see a list of all Products from first initialization of provider. So, is there any way to print all alive providers ? I do not want to initialize this provider with null clientUid in my second widget and print products using this initialization. 

Second method could be to print all products from first widget after a provider is disposed after choosing a client. How?

Third method could be someone with strong riverpod concepts suggests that what I explained is right and I should not worry about extra firebase costs here.

**Comment by sprabhdeep** [comment]
Here is full console output on first run of this app widget. Note that it prints twice with same number of items in the products list and hash value of provider is same (I'm using ProviderObserver class to print provider updates as per riverpod docs).

> Provider chosenClientUidProvider:AutoDisposeNotifierProviderImpl<ChosenClientUid, String?>#89929 was initialized with
> Provider fetchClientsProvider:AutoDisposeFutureProvider<List<User>>#e2535 was initialized with AsyncLoading<List<User>>()
> Provider fetchProductsProvider:FetchProductsProvider#aaaa9(null) was initialized with AsyncLoading<List<Product>>()
> Provider fetchProductsProvider:FetchProductsProvider#aaaa9(null) was initialized with AsyncLoading<List<Product>>()
> Provider fetchClientsProvider:AutoDisposeFutureProvider<List<User>>#e2535 updated from AsyncLoading<List<User>>() to
> AsyncData<List<User>>(value: [Instance of 'User', Instance of 'User'])
> Provider fetchProductsProvider:FetchProductsProvider#aaaa9(null) updated from AsyncLoading<List<Product>>() to AsyncData<List<Product>>(value: [Instance of
> 'Product', Instance of 'Product', Instance of 'Product', Instance of 'Product', Instance of 'Product', Instance of 'Product'])
> Provider fetchProductsProvider:FetchProductsProvider#aaaa9(null) updated from AsyncLoading<List<Product>>() to AsyncData<List<Product>>(value: [Instance of
> 'Product', Instance of 'Product', Instance of 'Product', Instance of 'Product', Instance of 'Product', Instance of 'Product'])

// After choosing a client from DropDownMenu

> Provider chosenClientUidProvider:AutoDisposeNotifierProviderImpl<ChosenClientUid, String?>#89929 updated from  to 1pj7MzqvnJWTFabcdef5vbe0rlq2
> Provider fetchProductsProvider:FetchProductsProvider#2de51(null) was initialized with AsyncLoading<List<Product>>()
> Provider fetchProductsProvider:FetchProductsProvider#aaaa9(null) was disposed
> Provider fetchProductsProvider:FetchProductsProvider#2de51(null) updated from AsyncLoading<List<Product>>() to AsyncData<List<Product>>(value: [Instance of
> 'Product', Instance of 'Product', Instance of 'Product'])
